### PR TITLE
Make extension install command find compatible version

### PR DESCRIPTION
**Pull request by vidartf**
Currently, the following pattern can occur when installing an extension:
 - Somebody has a version of jupyerlab which is not the latest version.
 - They try to install an extension, whose latest version depends on a newer version of the jupyterlab packages than what is installed.
 - They receive compatibility errors in the output, and either give up, ask for help, or if they know enough about lab extension and npm, figure out which version is compatible, and install that.

See e.g. #3215.

This PR will resolve this issue by the following steps:
 - When a package fails to install because of compatibility issues it:
   - Fetches the package metadata from the npm registry.
   - Finds the latest version of the package that is not incompatible, and install that.
   - If no such package can be found, fail as before.

### CODE REVIEW

**Comment by vidartf** [comment]
TODO: This should probably have a few tests, but I'm not entirely sure how to best mock the registry response.

File: jupyterlab/commands.py [hunk 1]
```diff
@@ -315,6 +317,8 @@ def __init__(self, app_dir, logger=None, kill_event=None):
         self.logger = logger or logging.getLogger('jupyterlab')
         self.info = self._get_app_info()
         self.kill_event = kill_event or Event()
+        # TODO: Make this configurable
+        self.registry = 'https://registry.npmjs.org'
```
Comment (blink1073) [inline jupyterlab/commands.py]: This is in the `lerna.json` file in the repo root.

Comment (blink1073) [reply jupyterlab/commands.py]: Actually that wouldn't help is because it isn't shipped. Will think about this more...

Comment (blink1073) [reply jupyterlab/commands.py]: Yeah I think leaving it as is and making it part of the extension application config sounds good

**Comment by blink1073** [comment]
As for testing, we could purposely ask to install an older version of an extension and make sure it successfully picks up the current one.  The tests are already reaching out to npm.

**Comment by vidartf** [comment]
> we could purposely ask to install an older version of an extension and make sure it successfully picks up the current one

That won't work due to the `'@' not in extension` check [here](https://github.com/jupyterlab/jupyterlab/pull/3608/files#diff-f889c27ebc01f0305bbef39cfd65085bR1107). Neither would I expect that behavior if I ask for a specific version of an extension.

**Comment by blink1073** [comment]
Ah, I see what you mean about the specific version, but I don't understand the reason for the `@` check.

**Comment by blink1073** [comment]
Is that meant to catch a version specifier?  If so, it should be checking after the first character to handle scoped packages.

**Comment by vidartf** [comment]
Ah, yes. It should get a tweak for that :)

**Comment by blink1073** [comment]
We should update the initial comment in `semver.py` to state that it is a modified version of the source.

**Comment by blink1073** [comment]
As far as testing, you could take a known package, pull the metadata associated with it in the mock, and then add a dummy newest version that depends on `@jupyterlab/application@1000` or something outrageous.

**Comment by vidartf** [comment]
I'm realized I had to deal with some nuances regarding the semver:
 - Dealt with non-standard preprelease strings.
 - Probably the user do not want to use prereleases at all actually, so maybe just filter these out? Or maybe we should sort them last so they are only considered if no stable release works?

**Comment by blink1073** [comment]
👍  to using prereleases last

**Comment by vidartf** [comment]
I refactored and rebased:
 - Only changes to semver version is to update to latest master (is there a reason why this is vendored and not imported?)
 - The key function is now in commands and able to accommodate sorting pre-releases last.

**Comment by blink1073** [comment]
Looks great, thanks!  I'm going to wait to merge until master is back to version 0.32.0dev.

**Comment by vidartf** [comment]
> - If no such package can be found, fail as before.

Maybe this behavior should be changed to: If no such package can be found, give a descriptive error saying: "No version of {package_name} could be found that is compatible with the current version of JupyterLab. The conflicting dependencies are detailed below." And then print the compatibility table like now.

**Comment by vidartf** [comment]
PS: It could also add a note about the nearest compatible matches, e.g.:
- If the newest supported version of lab is less than the current version, the message could say: "This extension does not yet support JupyterLab version X. The last version of JupyterLab that the extension supports is version Y".
- If the current version of lab is lower than the first supported version of the extension, it could say: "This extension does not support JupyterLab versions older than version X. Consider upgrading your JupyterLab installation".
- If the current version of lab is between two versions that the extension supports, it could say: "This extension does not support JupyterLab version X. The next supported version is Y, and the previous supported version is Z". (alternatively, it could instead end with "These are the versions of lab that the extension supports:" And then print a compatibility table).

**Comment by blink1073** [comment]
I think we should go with the better error message and strive towards 1.0 versions of the JS packages in the next release, which will help enormously with compatibility issues.

**Comment by vidartf** [comment]
I added some more commits to:
- Improve network error handling
- Fix some issues from initial work here
- Add more helpful error messages (not as detailed as outlined above)

**Comment by vidartf** [comment]
I have no idea why CI is failing here.

**Comment by blink1073** [comment]
This is great, thanks!

**Pull request merged**